### ISSUE TITLE
Attach to extension host

### ISSUE THREAD

**Issue by roblourens** [Microsoft/vscode-node-debug2#4]
Not sure how this is different from attaching to a typical node process

**Comment by roblourens** [comment]
Blocked? https://github.com/electron/electron/issues/6634

**Comment by roblourens** [comment]
Blocked on https://github.com/Microsoft/vscode/issues/28582

**Comment by weinand** [comment]
@roblourens I'll provide detailed info for what needs to be done here.

**Comment by weinand** [comment]
Here are some detailed steps for what needs to be done to support "inspector" protocol based EH debugging:

- I've added a startSession command for debug type "extensionHost" [here](https://github.com/Microsoft/vscode-node-debug/blob/master/src/node/extension/extension.ts#L243) which allows to use `"protocol": "inspector"` to switch the debug type to "extensionHost2". In the final version we don't need the switch because we will hardcode this (because we know the Electron version).
- node-debug2 needs to introduce a debug type "extensionHost2" similar to "node2".  There is no need to specify an attribute schema because "extensionHost2" will not be used in launch.configs (and there is no schema validation based on it).
- the "extensionHost2" type will be used in the node-debug2 implementation to do things a bit differently in three places.
  - in `launchRequest` two additional arguments are added: a [debug port and a debug session ID](https://github.com/Microsoft/vscode-node-debug/blob/master/src/node/nodeDebug.ts#L880-L884). Then the runtimeExecutable is spawned with all arguments but the DA **does not** attach because VS Code sends an explicit `attachRequest` later. This is different from the regular launch.
  - in the `attachRequest` [we fix a mode flag to be `launched`](https://github.com/Microsoft/vscode-node-debug/blob/master/src/node/nodeDebug.ts#L1237). If you don't have such a flag, you don't have to do anything.
  - in the `disconnectRequest` we [try to detect whether the terminate is part of a "restart"](https://github.com/Microsoft/vscode-node-debug/blob/master/src/node/nodeDebug.ts#L1536). In this case we do not kill the debuggee because VS Code does this in a better way.
- Optionally you can work around the issue that some "rejects" in the startup phase of the EH result in the debugger breaking on reject exceptions. This is painful if it happens with every reload of the EH. node-debug has introduced a third exception configuration option for this. We enable the existence of the option in the [`initializeRequest `](https://github.com/Microsoft/vscode-node-debug/blob/master/src/node/nodeDebug.ts#L750).

**Comment by roblourens** [comment]
Thanks! Will start on this today.

**Comment by roblourens** [comment]
I implemented this but there are two issues: when the attach request comes in, the debug config has type 'extensionHost' instead of 'extensionHost2'. But that's not a big deal.

Also I can't get 'restart' to work. It looks like it's vscode's job to restart the extensionHost window? I have the fix to not kill the process for a restart, and I send a TerminatedEvent the same as vscode-node-debug, but it doesn't reload the window. Debugged the vscode side but I'm not sure what's missing. cc @isidorn

**Comment by weinand** [comment]
@isidorn could you send the "attach" request with the massaged config that you receive in the 'vscode.startDebug' command?
I'm converting 'extensionHost' to 'extensionHost2' for Electron 1.7.4?

@roblourens Isi has fixed "restart" in today's Insiders by basically doing a reload window on the extension host window. Did you try with that version? Without the fix I was not seeing any effect with node-debug when pressing the restart button. That was bug Microsoft/vscode#31445.

**Comment by roblourens** [comment]
Thanks for the tip, I'm building out of the electron/1.7.4 branch so I'll try with newer stuff merged in.

**Comment by weinand** [comment]
Yes, if the electron branch hasn't been updated, it misses several fixes for EH debugging problems (that got introduced because of the multi-folder work).

**Comment by isidorn** [comment]
@weinand could you please elaborate your question. Currently the only place where I call attach is [here](https://github.com/Microsoft/vscode/blob/master/src/vs/workbench/parts/debug/electron-browser/debugService.ts#L837) and before that the configurations is massaged.

**Comment by weinand** [comment]
@isidorn in EH debugging, the launchRequest starts the extension host (via VS Code) but the DA does not connect to it (in normal "node" debugging the DA connects to the launched debuggee automatically).

Instead VS Code knows when the EH is ready and then broadcasts an event that you will pick up and issue the missing "attach" request for it. This is basically the same sequence as the "reload EH window" sequence.

For the "attach" request that you issue in response to the broadcast, you use the initial "launch" configuration and just change the "request" attribute from "launch" to "attach". 
Since the initial "launch" config uses the type "extensionHost", the "attach" uses that too.

The fix is to remember the massaged launch config (that uses the type "extensionHost2"), and create the "attach" config based on this.

**Comment by isidorn** [comment]
@weinand thanks for the explanation, do we need this for july? If not, I will create a new issue for me

**Comment by weinand** [comment]
@isidorn No, not for July (and as Rob said: this is not a big deal).

We should fix this more for consistency: if a launch config is modified in the extension, we should always use the modified config and not fall back to the original config.

**Comment by isidorn** [comment]
@weinand makes sense, created https://github.com/Microsoft/vscode/issues/31659
